proposals/2024/12-27_dynamic-ruleset.md

```diff
@@ -0,0 +1,197 @@
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+<!-- Answer here -->
+
+### Database
+
+N/A
+
+### UI
+
+N/A
+
+### API
+
+N/A
+
+### CLI
+
+N/A
+
+### YAML
+
+A `evaluation` key will be added under `ruleset` for the new dynamic functionality.
+
+This will instruct Vela to execute the step if the `evaluation` expression returns `true`.
+
+Example pipeline:
+
+```yaml
+steps:
+  - name: simple-match
+    ruleset:
+      evaluation: "VELA_BUILD_AUTHOR == 'JordanSussman'"
+    commands:
+      - echo "this will only run if the build author name is JordanSussman"
+
+  - name: function-match
+    ruleset:
+      evaluation: "hasPrefix(VELA_BUILD_AUTHOR, 'Jordan')"
+    commands:
+      - echo "this will only run if the build author name has the prefix of Jordan"
+      
+  - name: new-and-old
+    ruleset:
+      evaluation: "VELA_BUILD_AUTHOR == 'JordanSussman'"
+      event: [tag]
+    commands:
+      - echo "this will only run if JordanSussman triggered the tag event"
+```
+
+## Implementation
+
+<!--
+This section is intended to explain how the solution will be implemented for the proposal.
+
+NOTE: If there are no current plans for implementation, please leave this section blank.
+-->
+
+Users will be able to use [expr](https://expr-lang.org/) syntax within the `evaluation` key to filter steps.
+
+Expr describes itself as:
+
+> Expr is a Go-centric expression language designed to deliver dynamic configurations with unparalleled accuracy, safety, and speed. Expr combines simple syntax with powerful features for ease of use.
+
+All of the [built-in Vela environment variables](https://go-vela.github.io/docs/reference/environment/variables/) will be available within the `evaluation` key to perform logic with.
+
+The server code changes will involve updating the existing ruleset match function to consider both the `ruleset` and `evaluation` keys when determining if the step should execute. Both `ruleset` and `evaluation` must return true for the step to proceed.
+
+**Please briefly answer the following questions:**
+
+1. Is this something you plan to implement yourself?
+
+<!-- Answer here -->
+
+Yes, Cargill will add the code required to deliver on this feature
+
+2. What's the estimated time to completion?
+
+<!-- Answer here -->
+
+The code changes required to incorporate this should be relatively minor, so the work involved shouldn't take too long. However, the timeline for when we will start is still to be determined.
+
+**Please provide all tasks (gists, issues, pull requests, etc.) completed to implement the design:**
+
+<!-- Answer here -->
+
+## Questions
+
+1. Should we name the new key `evaluation` or choose a different name?
+2. Should the new key be nested under `ruleset` or be at the same level?
+3. Should it be possible to use both the existing `ruleset` and the new `evaluation` feature simultaneously? If so, should both need to return true for the step to execute?
```

Comment:
Yeah I think it should be possible to use both. Something like
```yaml
ruleset:
  event: push
  branch: main
  evaluation: "hasPrefix(VELA_BUILD_AUTHOR, 'Jordan')"
```

Of course you could add the other two rules into the evaluation, but that feels a bit messy. So I like being able to declare them separately as they normally are.